src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -580,67 +527,65 @@ sealed private[immutable] trait LazyListFactory[+CC[+X] <: LinearSeq[X] with Laz
   * @define Coll `LazyList`
   */
 @SerialVersionUID(3L)
-object LazyList extends LazyListFactory[LazyList] {
-
-  protected def newCons[T](hd: => T, tl: => LazyList[T]): LazyList[T] = new LazyList.Cons(hd, tl)
+object LazyList extends SeqFactory[LazyList] {
+  private[this] val _empty = newLL(State.Empty)
 
   @SerialVersionUID(3L)
-  object Empty extends LazyList[Nothing] {
-    override def isEmpty: Boolean = true
-    override def head: Nothing = throw new NoSuchElementException("head of empty lazy list")
-    override def tail: LazyList[Nothing] = throw new UnsupportedOperationException("tail of empty lazy list")
-    def force: this.type = this
-    override def knownSize: Int = 0
-    override def iterator: Iterator[Nothing] = Iterator.empty
-    protected def tailDefined: Boolean = false
-    protected def headDefined: Boolean = false
+  private sealed trait State[+A] extends Serializable {
+    def head: A
+    def tail: LazyList[A]
+    def headDefined: Boolean
   }
 
-  @SerialVersionUID(3L)
-  final class Cons[A](hd: => A, tl: => LazyList[A]) extends LazyList[A] {
-    private[this] var hdEvaluated: Boolean = false
-    private[this] var tlEvaluated: Boolean = false
-    override def isEmpty: Boolean = false
-    override lazy val head: A = {
-      hdEvaluated = true
-      hd
-    }
-    override lazy val tail: LazyList[A] = {
-      tlEvaluated = true
-      tl
+  private object State {
+    @SerialVersionUID(3L)
+    object Empty extends State[Nothing] {
+      def head: Nothing = throw new NoSuchElementException("head of empty lazy list")
+      def tail: LazyList[Nothing] = throw new UnsupportedOperationException("tail of empty lazy list")
+      def headDefined: Boolean = false
     }
-    def force: this.type = {
-      // Use standard 2x 1x iterator trick for cycle detection ("those" is slow one)
-      var these, those: LazyList[A] = this
-      if (!these.isEmpty) {
-        these.head
-        these = these.tail
-      }
-      while (those ne these) {
-        if (these.isEmpty) return this
-        these.head
-        these = these.tail
-        if (these.isEmpty) return this
-        these.head
-        these = these.tail
-        if (these eq those) return this
-        those = those.tail
+
+    @SerialVersionUID(3L)
+    final class Cons[A](private[this] var hd: () => A, val tail: LazyList[A]) extends State[A] {
+      @volatile private[this] var hdEvaluated: Boolean = false
+
+      lazy val head: A = {
+        hdEvaluated = true
+        val res = hd()
```

Comment:
Should the two lines above be swapped (in case evaluating `hd` throws)?

Reply:
It's probably better to keep retrying instead of switching from an exception to nulls.

Reply:
@Ichoran if it throws, `hd` won't get set to `null`, so it will keep retrying